### ISSUE TITLE
[Beta 8 blocker] CoreCLR on Mac crashes hard when there's a hardware exception

### ISSUE THREAD

**Issue by muratg** [dotnet/coreclr#1700]
The program pasted below will crash hard when attempting to access “test.Length”. The catch block will not execute. The exit code is consistently 6.

Simple repro:

```csharp
using System;

namespace Test
{
    public class Program
    {
        public void Main(string[] args)
        {
            string test = null;
            Console.WriteLine("NullReferenceException incoming?");
            
            try
            {
                Console.WriteLine("Hello {0}", test.Length);
            }
            catch (Exception ex)
            {
                Console.WriteLine("test.Length threw: {0}", ex);
                throw;
            }
        }
    }
}
```

System:
```
> dnvm list

Active Version              Runtime Arch OperatingSystem Alias
------ -------              ------- ---- --------------- -----
  *    1.0.0-beta8-15836    coreclr x64  darwin          default


> dnu list

[Target framework DNXCore,Version=v5.0 (dnxcore50)]

Package references:
  System.Console 4.0.0-beta-23406
  System.IO 4.0.0
  System.Runtime 4.0.0
  System.Text.Encoding 4.0.0
  System.Threading.Tasks 4.0.0
```

cc @joshfree @janvorli @sergiy-k @Eilon

**Comment by mellinoe** [comment]
I also hit this consistently in some testing yesterday. Even doing something as simple as:

````CSharp
Console.WriteLine(((object)null).ToString());
````
would crash the process (no error message whatsoever). In some of my real tests which had more complicated exceptions/flow, I was getting similar crashes, with an "Abort trap: 6" message (seems to be what you mentioned above).

I had assumed it was in some way related to the recent exception handling changes, and it looked like there were several PR's out related to exception handling.

**Comment by muratg** [comment]
BTW, this is OS X only.  On Linux this is fine.

**Comment by mellinoe** [comment]
I'm still seeing this issue, and I just pulled down commit b2d00c37749 and re-built.

**Comment by janvorli** [comment]
That commit is not the fix itself. The fix is #1699 and the #1701 is a missing file that belongs to the fix.

**Comment by mellinoe** [comment]
I see, but those PR's were already merged in prior to b2d00c3, right? I just built from HEAD of dotnet/coreclr master and the issue is still repro'ing for me. Unless I misunderstood what you meant. Should this be fixed in master now?

**Comment by janvorli** [comment]
@mellinoe Yes, it should be fixed in master. Let me try to build your test - I have a similar one that passes ok and was crashing before.

**Comment by mikem8361** [comment]
@mellinoe, I just ran your test and it seem to work now:

./corerun NullException.exe
NullReferenceException incoming?
test.Length threw: System.NullReferenceException: Object reference not set to an instance of an object.
   at Test.Program.Main(String[] args)

Unhandled Exception: System.NullReferenceException: Object reference not set to an instance of an object.
   at Test.Program.Main(String[] args)

**Comment by janvorli** [comment]
@mellinoe The same here.

**Comment by mellinoe** [comment]
Okay great, thanks for double checking. I'll try cleaning out my whole test directory to make sure nothing else was impacting this.

**Comment by mellinoe** [comment]
Hm, after re-building everything, I get a different error:

````
libc++abi.dylib: terminating with uncaught exception of type PAL_SEHException
Abort trap: 6
````

**Comment by janvorli** [comment]
@mellinoe Could you please run the test under lldb and paste here the stack when the issue happens?
If you are not familiar with lldb, you can just prepend "lldb -- " before the corerun command and when it hits the issue in the debugger, use "bt" command to get the stack trace.

**Comment by mellinoe** [comment]
Here's all the output from the run:

````
osx-coreclr erme$ lldb -- ./corerun NetworkInformation.SampleUsage.dll 
(lldb) target create "./corerun"
Current executable set to './corerun' (x86_64).
(lldb) settings set -- target.run-args  "NetworkInformation.SampleUsage.dll"
(lldb) run
Process 79118 launched: './corerun' (x86_64)
libc++abi.dylib: terminating with uncaught exception of type PAL_SEHException
Process 79118 stopped
* thread #1: tid = 0xe9109, 0x00007fff905bc286 libsystem_kernel.dylib`__pthread_kill + 10, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x00007fff905bc286 libsystem_kernel.dylib`__pthread_kill + 10
libsystem_kernel.dylib`__pthread_kill:
->  0x7fff905bc286 <+10>: jae    0x7fff905bc290            ; <+20>
    0x7fff905bc288 <+12>: movq   %rax, %rdi
    0x7fff905bc28b <+15>: jmp    0x7fff905b7c53            ; cerror_nocancel
    0x7fff905bc290 <+20>: retq   
(lldb) bt
* thread #1: tid = 0xe9109, 0x00007fff905bc286 libsystem_kernel.dylib`__pthread_kill + 10, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
  * frame #0: 0x00007fff905bc286 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff8db719f9 libsystem_pthread.dylib`pthread_kill + 90
    frame #2: 0x00007fff95df59b3 libsystem_c.dylib`abort + 129
    frame #3: 0x00007fff94466a21 libc++abi.dylib`abort_message + 257
    frame #4: 0x00007fff9448e9d1 libc++abi.dylib`default_terminate_handler() + 267
    frame #5: 0x00007fff8fe2b7eb libobjc.A.dylib`_objc_terminate() + 124
    frame #6: 0x00007fff9448c0a1 libc++abi.dylib`std::__terminate(void (*)()) + 8
    frame #7: 0x00007fff9448bd48 libc++abi.dylib`__cxa_rethrow + 99
    frame #8: 0x000000010134f5b1 libcoreclr.dylib`RaiseTheExceptionInternalOnly(throwable=OBJECTREF @ 0x00007fff5fbfee18, rethrow=NO, fForStackOverflow=NO) + 2305 at excep.cpp:3115
    frame #9: 0x0000000101351949 libcoreclr.dylib`UnwindAndContinueRethrowHelperAfterCatch(pEntryFrame=0xffffffffffffffff, pException=0x000000011bf094a0) + 601 at excep.cpp:8706
    frame #10: 0x00000001012f6393 libcoreclr.dylib`CorHost2::ExecuteAssembly(this=0x00000001005007d0, dwAppDomainId=1, pwzAssemblyPath=u"/Users/erme/osx-coreclr/NetworkInformation.SampleUsage.dll", argc=0, argv=0x0000000000000000, pReturnValue=0x00007fff5fbff31c) + 1571 at corhost.cpp:1304
    frame #11: 0x0000000101134315 libcoreclr.dylib`::coreclr_execute_assembly(hostHandle=0x00000001005007d0, domainId=1, argc=0, argv=0x0000000000000000, managedAssemblyPath="/Users/erme/osx-coreclr/NetworkInformation.SampleUsage.dll", exitCode=0x00007fff5fbff31c) + 341 at unixinterface.cpp:286
    frame #12: 0x0000000100003396 corerun`ExecuteManagedAssembly(currentExeAbsolutePath="/Users/erme/osx-coreclr/corerun", clrFilesAbsolutePath="/Users/erme/osx-coreclr", managedAssemblyAbsolutePath="/Users/erme/osx-coreclr/NetworkInformation.SampleUsage.dll", managedAssemblyArgc=0, managedAssemblyArgv=0x0000000000000000) + 3542 at coreruncommon.cpp:311
    frame #13: 0x000000010000743c corerun`corerun(argc=2, argv=0x00007fff5fbffc20) + 2284 at corerun.cpp:150
    frame #14: 0x0000000100007502 corerun`main(argc=2, argv=0x00007fff5fbffc20) + 34 at corerun.cpp:162
    frame #15: 0x00007fff948c55c9 libdyld.dylib`start + 1
````

**Comment by janvorli** [comment]
I was able to repro the issue. When I run corerun without sudo, it happens for me too. I was testing it under lldb and I run that always with sudo, so I haven't noticed that.
I am looking into it.

**Comment by janvorli** [comment]
In my case, the issue was that the managed exe was owned by root and not by me, so that was causing the issue. chown to myself fixed the problem.
But it obviously uncovered an issue in the new exception handling - we are missing a high level exception trap in the ExecuteAssembly and so the native exception leaks. I have put the handler back and now it displays correctly the exception:

`Unhandled Exception: System.IO.FileLoadException: Could not load file or assembly '/Users/janvorli/xunit/nullref2.exe' or one of its dependencies. Access is denied.`

I'll send a PR soon for a fix.

**Comment by janvorli** [comment]
The fix is in as #1715

**Comment by joshfree** [comment]
Closing the Beta8 bug since @janvorli 's fix is in the 23408.00 release build (published last night to the aspnetbeta8 test feed)

**Comment by mellinoe** [comment]
> Closing the Beta8 bug since @janvorli 's fix is in the 23408.00 release build (published last night to the aspnetbeta8 test feed)

I'm not sure I follow. The fix in #1715 was just checked in an hour or so ago. I also haven't had a chance to verify that my issue is fixed, either.

**Comment by mellinoe** [comment]
Unfortunately, I'm still hitting an issue, although not the exact same one. I will file another issue.

**Comment by joshfree** [comment]
To clarify, the two fixes https://github.com/dotnet/coreclr/pull/1699 and https://github.com/dotnet/coreclr/pull/1701 were pulled into the 1.0.0-beta8 release branch on VSO and verified by ASP.NET as to unblock Kestrel on Mac.  Other bugs should be filed separately for 1.0.0-rc1.
